src/servers/src/grpc/authorize.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::pin::Pin;
+use std::result::Result as StdResult;
+use std::task::{Context, Poll};
+
+use auth::UserProviderRef;
+use hyper::Body;
+use session::context::QueryContext;
+use tonic::body::BoxBody;
+use tonic::transport::NamedService;
+use tower::{Layer, Service};
+
+use crate::http::authorize::{extract_catalog_and_schema, extract_username_and_password};
+
+#[derive(Clone)]
+pub struct AuthMiddlewareLayer {
+    user_provider: Option<UserProviderRef>,
+}
+
+impl AuthMiddlewareLayer {
+    pub fn with(user_provider: Option<UserProviderRef>) -> Self {
+        Self { user_provider }
+    }
+}
+
+impl<S> Layer<S> for AuthMiddlewareLayer {
+    type Service = AuthMiddleware<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        AuthMiddleware {
+            inner: service,
+            user_provider: self.user_provider.clone(),
+        }
+    }
+}
+
+/// This middleware is responsible for authenticating the user and setting the user
+/// info in the request extension.
+///
+/// Detail: Authorization information is passed in through the Authorization request
+/// header.
+#[derive(Clone)]
+pub struct AuthMiddleware<S> {
+    inner: S,
+    user_provider: Option<UserProviderRef>,
+}
+
+impl<S> NamedService for AuthMiddleware<S>
+where
+    S: NamedService,
+{
+    const NAME: &'static str = S::NAME;
+}
+
+type BoxFuture<'a, T> = Pin<Box<dyn std::future::Future<Output = T> + Send + 'a>>;
+
+impl<S> Service<hyper::Request<Body>> for AuthMiddleware<S>
+where
+    S: Service<hyper::Request<Body>, Response = hyper::Response<BoxBody>> + Clone + Send + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxFuture<'static, StdResult<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<StdResult<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: hyper::Request<Body>) -> Self::Future {
+        let clone = self.inner.clone();
+        let mut inner = std::mem::replace(&mut self.inner, clone);
```

Comment:
Can you simply use the cloned "inner"?

```suggestion
        let mut inner = self.inner.clone();
```

Reply:
I need to add some comments to explain it.